orchestrator: Reconcile instances, not tasks

aaronlehmann opened the pull request:
This commit prepares for an upcoming change that will allow a
configurable rollout strategy (either shut down the old task before
starting the new one, or start the new one before shutting down the old
one). The issue that was blocking having a rollout strategy with an
overlap between the lifetime of the new and old tasks was that
service-level reconciliation couldn't properly handle situations where
there was more than one active task per slot or node. This change
attempts to solve this by moving to *instance-level* reconciliation.

Before, the orchestrator counted tasks. With this change, it counts
instances (slots or nodes) that are satisfied with at least one task
that has a desired state > `NEW` and <= `RUNNING`.

When service-level reconciliation calls the updater, it passes
`[][]*api.Task` instead of a flat `[]*api.Task`. The first dimension is the
instance (slot or node grouping), and the next dimension contains the
set of tasks for that instance. This allows the updater to handle
situations like partial updates. For example, if slot 5 has two running
tasks, and only one matches the current service spec, the updater can
recognize that this is a partial update, and remove the dirty task. In
this model, the orchestrator doesn't need to know anything about what's
going on with the individual tasks. It just has scale up and down as
appropriate, and let the updater reconcile state below the instance
level.

I explored a few alternative approaches as well, and ran into a lot of
complexity because the orchestrator invariably had to recognize partial
update situations, and take action to remediate them (making sure not to
fight with a current update in progress). I find this model more elegant
because the orchestrator only needs to call Update, as it did before,
and the updater handles all logic related to updates. But I'm interested
in feedback and I'm willing to explore more alternatives.


cc @dongluochen @aluzzardi

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "rollout-strategy" git@github.com:aaronlehmann/swarmkit.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "rollout-strategy" git@github.com:aaronlehmann/swarmkit.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "rollout-strategy" git@github.com:aaronlehmann/swarmkit.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "rollout-strategy" git@github.com:aaronlehmann/swarmkit.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

dongluochen commented on manager/orchestrator/updater.go:
```diff
@@ -302,6 +389,10 @@ func (u *Updater) isServiceDirty(service *api.Service) bool {
 }
 
+func (u *Updater) isInstanceDirty(instance []*api.Task) bool {
+	return len(instance) > 1 || u.isTaskDirty(instance[0])
```
`len(instance) > 1 || (len(instance) == 1 && u.isTaskDirty(instance[0]))`
I think it would not happen in this context. But I feel it better to make sure no out of bound access.

dongluochen commented:
LGTM

aaronlehmann commented:
ping @aluzzardi - I would like your agreement on this one before merging it. It's going to influence all of the rolling update changes that I'm working on. I'm somewhat blocked on making rolling update changes until I know whether I should make them on top of this updater design, or some other approach.

aluzzardi commented:
nit: Wondering if we should have a few high level types, e.g. `type slot []*api.Task` / `type slots []slot` with helper member functions so we can concentrate some of that logic into a single place.

WDYT?

Some mild concern about the number of for loops on what can be quite a large number. Not entirely related to this PR as the problem existed before, it just makes it slightly worse.

Having higher level objects might help with caching/precomputing as well (e.g. running tasks, "active" tasks, and what not) - or at the very least, it'd help isolate expensive work in pprof.

Regarding terminology, shouldn't `instances` be named `slots` and `instance` a `slot`?

Unrelated: At the cost of sounding like a broken record, we do still pay a small price for not using slots for global services every single time we make changes around this part of the code.

Overall LGTM

aaronlehmann commented:
> nit: Wondering if we should have a few high level types, e.g. type slot []*api.Task / type slots []slot with helper member functions so we can concentrate some of that logic into a single place.

I like this idea. I'll see what I can do.

> Regarding terminology, shouldn't instances be named slots and instance a slot?

Yeah, I think `Slot` used to be called `Instance` but the code that dealt with these was never updated, and I'm propagating the old name here. I'll fix it.

aaronlehmann commented:
> Yeah, I think `Slot` used to be called `Instance` but the code that dealt with these was never updated, and I'm propagating the old name here. I'll fix it.

Actually, now that I think about it, there's the fact that global services don't use slots (which I know you dislike).

So `Instance` was partially an attempt to have a more generic term that refers to an instance of the service, not necessarily a slot.

But I can change the variable names to `slot` and `slots` if you think it makes sense even though the underlying organization may not actually be based on slot numbers.

aluzzardi commented:
Maybe it'd be easier to go with slots anyway even if the terminology wouldn't be 100% accurate all the time just to reduce the number of concepts we have around (services, slots, instances, tasks)

aluzzardi commented:
Oh, and replicas :)

aaronlehmann commented:
I renamed "instances" to "slots" and refactored this to use a `type slot []*api.Task`. I think it's cleaner than before. I also moved the slot sorting stuff into `slot.go` along with the type declaration.

I found some code I could factor out of `reconcile` and reuse in the unit tests. There's a new `getRunnableSlots` function that creates a map of slots for a replicated service.

I didn't find anything that really made sense to factor out as a `slot` method. The stuff the replicated orchestrator does with slots is pretty replicated-orchestrator-specific; the stuff the updater does with them (checking if specs are dirty, etc.) is pretty updater-specific. But let me know if you have any ideas for cleaning this up further.

aluzzardi commented:
LGTM
